packages/pyright-internal/src/analyzer/program.ts

```diff
@@ -240,6 +245,37 @@ export class Program {
         this._cacheManager.unregisterCacheOwner(this);
     }
 
+    enterEditMode() {
```

Comment:
It would probably be better to not expose `enterEditMode` and `exitEditMode` directly. These should be private. The public interface should be something like `useEditMode` with a try/catch handler, like this:

```typescript
useEditMode(callback: () => void) {
    this._enterEditMode();
    try {
        callback();
   } finally {
        this._leaveEditMode();
    }
}
```

This will properly handle cancellation exceptions.

Reply:
if possible, it would be safer if we define something like ProgramView, like ClonedProgram or something like that and allow only what ProgramView allow + updating only text but doesn't allow updating like configOption (such as setConfiguration and etc)

by the way, will it allow modifying files? asking since one of common task on advanced refactoring is adding/removing/renaming a file.